src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Infrastructure/Streams.cs

```diff
@@ -1,21 +1,76 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+using System.IO;
 using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
 
 namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
 {
     class Streams
     {
+        private FrameRequestStream _upgradedRequest;
+        private FrameResponseStream _upgradedResponse;
+
         public Streams(IFrameControl frameControl)
         {
             RequestBody = new FrameRequestStream();
             ResponseBody = new FrameResponseStream(frameControl);
-            DuplexStream = new FrameDuplexStream(RequestBody, ResponseBody);
         }
 
         public FrameRequestStream RequestBody { get; }
         public FrameResponseStream ResponseBody { get; }
-        public FrameDuplexStream DuplexStream { get; }
+
+        public Stream Upgrade()
+        {
+            _upgradedRequest = _upgradedRequest ?? RequestBody.Upgrade();
```

Comment:
I think asking the primary RequestBody stream to return an upgraded version of itself causes some grossness. I don't like either FrameRequestStream or FrameResponseStream having an Upgraded state or an Upgrade method.

I think that an upgradable request should have completely fake RequestBody and ResponseBodys that no-op. The upgraded stream should be built from the real streams that aren't accessible via the HttpContext directly.